src/tls_stream.rs

```diff
@@ -588,12 +622,71 @@ impl<S> TlsStream<S>
                     self.out_buf.set_position(0);
                     Ok(())
                 }
-                err => Err(io::Error::from_raw_os_error(err as i32)),
+                err => {
+                    let err = io::Error::from_raw_os_error(err as i32);
+                    debug!("failed to encrypt: {:?}", err);
+                    Err(err)
+                }
             }
         }
     }
 }
 
+impl<S> MidHandshakeTlsStream<S>
+    where S: Read + Write,
+{
+    /// Returns a shared reference to the inner stream.
+    pub fn get_ref(&self) -> &S {
+        self.inner.get_ref()
+    }
+
+    /// Returns a mutable reference to the inner stream.
+    pub fn get_mut(&mut self) -> &mut S {
+        self.inner.get_mut()
+    }
+
+    /// Restarts the handshake process.
+    pub fn handshake(mut self) -> Result<TlsStream<S>, HandshakeError<S>> {
+        match self.inner.initialize() {
+            Ok(_) => Ok(self.inner),
+            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                debug!("handshake incomplete");
+                Err(HandshakeError::Interrupted(self))
+            }
+            Err(e) => {
+                debug!("handshake error: {:?}", e);
+                Err(HandshakeError::Failure(e))
+            }
+        }
+    }
+}
+
+impl<S: fmt::Debug + Any> Error for HandshakeError<S> {
```

Comment:
Ah, nevermind, the Error comment above.

Can these impls be moved next to the type definition though?